README.md
# git-remote-aws [![PyPI version](https://badge.fury.io/py/git-remote-aws.svg)](https://badge.fury.io/py/git-remote-aws)

`git-remote-aws` is a [git remote helper](https://git-scm.com/docs/git-remote-helpers) for pulling data from an AWS account just like pulling from a Git remote.

Repository published on [Gitlab](https://gitlab.com/autofitcloud/git-remote-aws) and [Github](https://github.com/autofitcloud/git-remote-aws)

Website published at https://git-remote-aws.autofitcloud.com

Check the wishlist at the bottom of this readme for future plans.

Vote for features on the [git-remote-aws public trello board](https://trello.com/b/szNMyNYJ/git-remote-aws).


<i>News 2019-08-23:</i>
I had posted this project on [r/git](https://www.reddit.com/r/git/comments/ctxcq8/gitremoteaws_aws_accounts_as_git_remotes/) last night,
only to wake up this morning and find it on top of the subreddit!
r/git is awesome!
[Screenshot](https://imgur.com/gallery/4PB2BeY)

<i>More news 2019-08-23:</i>
Apparently my cross-post to [r/aws](https://www.reddit.com/r/aws/comments/cu2llv/gitremoteaws_aws_accounts_as_git_remotes/) also made it to #3 overnight! [Screenshot](https://imgur.com/gallery/ndtJ4wc)


## Installation

```
sudo apt-get install git python3 python3-pip

pip install awscli git-remote-aws
```

## Basic Usage

Configure `awscli` with your AWS key and secret (skip this step if already done)

```
aws configure
```

The role or user for the configuration should have a subset (or all) of the following policies attached

```
AmazonEC2ReadOnlyAccess
AWSCloudTrailReadOnlyAccess
CloudWatchReadOnlyAccess
```

Init a new git repo

```
TMPDIR=`mktemp -d`
cd $TMPDIR
git init
```

Add AWS remotes for EC2 describe-instances, list-metrics, etc.

```
git remote add ctle_ec2Typechanges      aws+cloudtrail::/lookup-events?filter=ec2TypeChanges
git remote add cw_descAlarms      aws+cw::/describe-alarms
git remote add cw_getMetricData   aws+cw::/get-metric-data
git remote add cw_listMetrics     aws+cw::/list-metrics
git remote add ec2_descInstances  aws+ec2::/describe-instances
git remote add sns_listTopics     aws+sns::/list-topics
```

Fetch data from all remotes.

```
git fetch --all
```

This creates a folder `aws.amazon.com` with a directory structure containing the relevant data

```
> tree
.
└── aws.amazon.com
    └── us-west-2
        └── ec2_describeInstances
        │   ├── i-02432bc7.json
        │   ├── i-069a7808addd143c7.json
        │   ├── i-08c802de5accc1e89.json
        │   ├── i-0e2662888859c5507.json
        │   ├── i-0fb05d874895a05ec.json
        │   ├── i-34ca2fc2.json
        │   └── i-e1ca46eb.json
        └── ...

4 directories, 11 files
```


## Advanced Usage

### awscli profiles

To use a profile from `~/.aws/credentials` other than the default,
append `?profile=<optional profile name to use>` to the remote URLs.


### boto3 options

To append other boto3 Session constructor arguments as documented
[here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html),
append `?boto3_session_config=path/to/file` to the remote URLs,
where `path/to/file` points to a JSON file containing the arguments from the boto3 session constructor.

For example,

```
{ "aws_access_key_id": "ABC", "aws_secret_access_key": "ABC", ...}
```


Note: The default behavior of the `describe-instances` endpoint is to subset the EC2 description to a minimal.
To get the full EC2 desriptions, append `?fulldata=true` to the endpoint.


### git push remotes

Push to a git remote

```
git add aws.amazon.com
git commit -m 'first commit'

git remote add origin git@gitlab.com:shadiakiki1986/shadiakiki1986.aws.amazon.com-json.git
git push -u origin master
```


## Covered services

The following AWS services are currently covered

Service    | Command | Notes
--------|------|------
EC2     | describe-instances | -
Cloudtrail | lookup-events | Custom filter of results for EC2 instance type changes.
Cloudwatch   | list-metrics | -
Cloudwatch   | get-metric-data | -
Cloudwatch   | describe-alarms | -
SNS | list-topics | -


Required policies/permissions per service

Service    | Policy
--------|------
EC2     | AmazonEC2ReadOnlyAccess
Cloudtrail | AWSCloudTrailReadOnlyAccess
Cloudwatch | CloudWatchReadOnlyAccess
SNS | -


### AWS/Moto endpoints

The full structure of the remote URLs is as follows

```
git remote add example_1 aws+<service>::<endpoint url>/<command>?profile=<optional profile name to use>&boto3_session_config=path/to/file
```

where

- `service` is one of: `ec2`, `cw` (cloudwatch), `sns`
- `endpoint url` is the AWS endpoint to use
    - leave blank for the default AWS endpoints, or use [moto](http://docs.getmoto.org/en/latest/) for mocked AWS services
- `command`: depending on the `service` above, this can be
    - `ec2`:
        - `describe-instances`
        - `catalog` (this is not an official AWS service, but is populated from https://www.ec2instances.info. Check related note in "Developer notes" below)
    - `cw`:
        - `list-metrics`
        - `get-metric-data`
        - `describe-alarms`
    - `sns`
        - `list-topics`
- `profile` is the profile name from `~/.aws/credentials`
    - Only one `profile` is supported at a time ATM, check [issue #5](https://gitlab.com/autofitcloud/git-remote-aws/issues/5)
- `boto3_session_config` is a json file of key-value pairs corresponding to boto3 session constructor arguments
    - this is optional
    - documentation for boto3 session is [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html)
    - Only one `boto3_session_config` is supported at a time ATM, check [issue #5](https://gitlab.com/autofitcloud/git-remote-aws/issues/5)


Examples

```
# get from AWS using the default profile in ~/.aws/credentials
git remote add example_1_ec2 aws+ec2::/describe-instances
git remote add example_1_catalog aws+ec2::/catalog
git remote add example_1_cwListMetrics aws+cw::/list-metrics

# Specific aws endpoint
git remote add example_3 aws+ec2::http://ec2.us-west-2.amazonaws.com/describe-instances

# use a specific profile and AWS default endpoints
git remote add example_2 aws+ec2::/describe-instances?profile_name=profile&boto3_session_config=path/to/file
```

Pull the data

```
git fetch example_1_ec2
git fetch example_1_catalog
```


## Developer notes

Check `DEVELOPER.md`


## Wishlist

Here is a list of features that were brought up from reddit [r/git](https://www.reddit.com/r/git/comments/ctxcq8/gitremoteaws_aws_accounts_as_git_remotes/exsu7bs/?context=3) and [r/aws](https://www.reddit.com/r/aws/comments/cu2llv/gitremoteaws_aws_accounts_as_git_remotes/)

- [Issue #1](https://github.com/autofitcloud/git-remote-aws/issues/1): Add push capability to the aws+ec2://describe-instances remote
- [Issue #2](https://github.com/autofitcloud/git-remote-aws/issues/2): git fetch should build a history when possible


## Support

I built `git-remote-aws` as part of the workflow behind [AutofitCloud](https://autofitcloud.com), the early-stage startup that I'm founding, seeking to cut cloud waste on our planet.

If you like `git-remote-aws` and would like to see it developed further,
please support me by signing up at https://autofitcloud.com

Over and out!

--[u/shadiakiki1986](https://www.reddit.com/user/shadiakiki1986)


CHANGELOG.md
Version latest (to be tagged as 0.5.2) (2019-09-10)

- ...


Version 0.5.{4,5} (2020-01-27)

- enh: split out code of AWS ec2 catalog to a class and added mapping to smaller type without accounting for family
- bugfix: use file caching of www.ec2isntances.info file
- enh: mapping to same performance but cheaper instances in the recent `t3c_smaller_familyNone` function


Version 0.5.3 (2019-09-10)

- bugfix: case of no cloudtrail data found to be handled properly instead of throwing a weird pandas error


Version 0.5.2 (2019-09-03)

- FEAT: bump version for issue with pypi rendering markdown with table


Version 0.5.1 (2019-08-30, 2019-09-03)

- FEAT: use RunInstance event name on top of ModifyInstanceAttribute to get the initial instance type
- ENH: ec2 type changes to be stored one csv file per instance
- BUGFIX: no longer passing "StartTime" to cloudtrail history of ec2 type changes
- ENH: factor out `pull.py` into `pull_{ec2,cloudwatch,sns}.py`


Version 0.5.0 (2019-08-30)

- FEAT: add cloutrail endpoint for lookup-events with custom filter for ec2 type changes


Version 0.4.0 (2019-08-15)

- FEAT: replace `profile_name` in git remote URL with `boto3_session_kwargs` (JSON containing `profile_name` as one of the keys)
- BUGFIX: bring back `profile_name` and replace `boto3_session_kwargs` with `boto3_session_config` (JSON file for shorter URLs)
- ENH: comment out downloading hourly CPU


Version 0.3.9 (2019-08-{08,14})

- ENH: add cases to test
- ENH: improve docs
- BUGFIX: check for locale before setting it (faced when deploying on AWS lambda)
- bugfix: cancel 0.3.8


Version 0.3.7 (2019-08-08)

- ENH: add feature `?fulldata=true` to the `ec2_describeInstances` endpoint
    - this followed from strong criticism on reddit in my post: [Is opensource infrastructure safe?](https://www.reddit.com/r/aws/comments/cn81my/is_opensource_infrastructure_safe/)


Version 0.3.6 (2019-08-07)

- ENH: moved path of aws catalog from gitRemoteAws.pull into dotman.fn.awsCat
- BUGFIX: refactor `df` to `df_pd` to distnguish from `df_json` in `pull`
- BUGFIX: flattening the `pricing` field from `www.ec2instance.info` was yielding NaN. Replaced with the average price of all regions.


Version 0.3.5 (2019-08-07)

- ENH: readme cleanup
- ENH: split out developer notes to separate file than README.md
- released on pypi


Version 0.3.4 (2019-08-07)

- ENH: add description in setup.py


Version 0.3.3 (2019-08-07)

- ENH: change aws data root to be aws.amazon.com instead of just aws
- ENH: improved readme


Version 0.3.2 (2019-08-06)

- ENH: add progress output (still pending including a percentage, check issue # 6) with tqdm
- BUGFIX: comment out logger.setLevel (should be based on the -v, check issue # 7)
- BUGFIX: increase maxresults in fetch aws+ec2 instance descriptions from 30 to 3000 + add warning when the limit is hit


Version 0.3.1 (2019-08-02)

- BUGFIX: fix wrong variable reference
- FEAT: implement the `profile_name` parameter


Version 0.3.0 (2019-07-22)

- FEAT: add sns list-topics


Version 0.2.1 (2019-07-18)

- FEAT: drop deprecated piece of code in pull.py that was moved into `cli_*.py`
- FEAT: use jmespath in filtering the cloudwatch describe alarms response
- FEAT: implement cloudwatch list-metrics remote fetch
- FEAT: implement cloudwatch get-metric-data
- ENH: move `cw_getMetricStatistics` out of ec2op to here for later integration (check issue #4)
- ENH: use package name "git-remote-aws" with dashes instead of "gitRemoteAws" in setup.py
- ENH: add locale at top of cli.py files
- ENH: ec2 catalog coming from non-aws source to go into separate folder than aws subrepo
- BUGIFX: include statistic in folder name of get-metric-data


Version 0.2.0 (2019-07-16)

- FEAT: change format of remote URL so that it allows setting different endpoints
- FEAT: replace `print` with `logger.{info,debug,...}` so that the text shows up when using git
- FEAT: add ec2 catalog support (factored out some common code)
- FEAT: add logging stream to stderr
- FEAT: implement proper git <-> helper command exchange so that git no longer exits with non-0 code
- ENH: factor out `cli` to `cli_core` and `Ec2Class` in preparation for new `aws+cw` protocol
- FEAT: factor out and add alternative command for cloudwatch
- FEAT: add cloudwatch/describe-alarms



Version 0.1.0 (2019-07-12)

- FEAT: initial files
- FEAT: first working version


gitRemoteAws/pull.py
# Copied from gitlab.com/shadiakiki1986/ec2optimizer
# and modified as needed

from botocore.exceptions import ClientError
# from .botoman import SessionMan
import importlib
import yaml
import os
import sys
import datetime as dt
import json
import requests
from cachecontrol import CacheControl
import pandas as pd
import numpy as np
import copy
from tqdm import tqdm

import logging
logger = logging.getLogger("git-remote-aws")

##############################
# from lambda_login.py and lambda_ec2_describeInstances.py

def postprocess_response(response_1):
    if 'Reservations' not in response_1:
        return
    
    for res_i, res_v in enumerate(response_1['Reservations']):
        for i_desc in res_v['Instances']:

            # put dict to ddb
            item = {
                # keys
                'InstanceId': i_desc['InstanceId'],

                # some fields duplicated out of the original response
        		'OwnerId': response_1['Reservations'][res_i]['OwnerId'], # eg '886436197218'
        		'ReservationId': response_1['Reservations'][res_i]['ReservationId'], # eg 'r-e5813ce8'

                # core info
                'InstanceDescription': i_desc
            }
            yield item


def json_serial(obj):
    """
    JSON serializer for objects not serializable by default json code
    http://stackoverflow.com/questions/11875770/ddg#22238613
    """
    if isinstance(obj, (dt.datetime, dt.date)):
        return obj.isoformat()
    raise TypeError ("Type %s not serializable" % type(obj))


def get_instDesc(fn, ec2, fulldata):
    """
    fn - dict of local file paths to use
    ec2 - boto3 ec2 client
    fulldata - boolean, True when all data is to be dumped into the output. False when only the minimal necessary data is needed.
    """
    MaxResults=3000 # FIXME <<<<<<<<<<<<<<<<<<<<<<<<<<<
    
    # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_instances
    response_1 = ec2.describe_instances(
        DryRun=False, # False, # True, # |
        MaxResults=MaxResults # 1 # 30
    )
    
    # https://stackoverflow.com/a/55749807/4126114
    # , ec2.meta.region_name
    response_2 = postprocess_response(response_1)
    
    
    nr = 0 # counter to check if we're hitting the max above     
    for ri in tqdm(response_2, desc="Pulling instance desriptions"):
        nr+=1
        if nr==MaxResults: logger.warning("MaxResults=%i attained."%MaxResults)
        
        # trim information if not "fulldata"
        if not fulldata:
            ri.pop('OwnerId', None)
            ri.pop('ReservationId', None)
            ri['InstanceDescription'] = { k: ri['InstanceDescription'][k]
                                          for k in ri['InstanceDescription'].keys()
                                          if k in ['CpuOptions', 'InstanceType', 'InstanceId', 'Tags', 'State']
                                        }

        # save instance description
        fn_temp = os.path.join(fn['ec2DescInst'], ri['InstanceId']+'.json')
        # logger.debug("save ec2 desc to %s"%fn_temp)
        with open(fn_temp, 'w') as fh:
            json.dump(ri, fh, default=json_serial, indent=4, sort_keys=True)


# use jmespath like awscli
# https://stackoverflow.com/a/57018780/4126114
# Example
#   >>> mydata
#   {'foo': {'bar': [{'name': 'one'}, {'name': 'two'}]}}
#   >>> jmespath.search('foo.bar[?name==`one`]', mydata)
#   [{'name': 'one'}]
import jmespath

def get_cwCore(fn_dir, cloudwatch, operation_name, dict_key, **kwargs):
    """
    cloudwatch - boto3 cloudwatch client
    """
    # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudwatch.html#CloudWatch.Paginator.DescribeAlarms
    PaginationConfig = {
        'MaxResults': 3000 # FIXME <<<<<<<<<<<<<<<<<<<<<<<<<<<
    }

    # List alarms of insufficient data through the pagination interface
    # https://github.com/boto/boto3/blob/90c03b3aff081e13f5a8dfca2f37afe978ee4809/docs/source/guide/cw-example-creating-alarms.rst#describe-alarms
    # dimensions and namespace parameters
    # https://github.com/boto/boto3/blob/90c03b3aff081e13f5a8dfca2f37afe978ee4809/docs/source/guide/cw-example-metrics.rst#example
    # paginator = cloudwatch.get_paginator(operation_name, Namespace='AWS/EC2', MaxResults=MaxResults)
    paginator = cloudwatch.get_paginator(operation_name)
    for i, response in tqdm(enumerate(paginator.paginate(PaginationConfig=PaginationConfig, **kwargs)), desc="Downloading %s"%operation_name):
        # filter for instance descriptions
        #sys.stderr.write(json.dumps(list(response.keys()), default=json_serial, indent=4, sort_keys=True))
        #sys.stderr.write(json.dumps(response, default=json_serial, indent=4, sort_keys=True))
        #sys.stderr.write('\n')
        response[dict_key] = jmespath.search(dict_key+'[?Namespace==`AWS/EC2`]', response)
        
        for j, operation_i in enumerate(response[dict_key]):
            #sys.stderr.write("%s/%s: %s\n"%(i, j, json.dumps(operation_i['Dimensions'])))
            #return # FIXME
        
            # save instance description
            # Note that this is another filtering step that maybe is unnecessary
            # because any "AWS/EC2" namespace item should have the "InstanceId" dimension
            # Keeping it anyway because this code is untested ATM <<<<<<<<<<<<<<<<<<<<<<<<<<<<<< FIXME
            iid = [x['Value'] for x in operation_i['Dimensions'] if x['Name']=='InstanceId']
            if len(iid)==0: continue
            
            # sys.stderr.write(json.dumps(operation_i, default=json_serial, indent=4, sort_keys=True))
            # sys.stderr.write('\n')
    
            iid = iid[0]
            fn_temp = os.path.join(fn_dir, iid+'.json')
            #logger.debug("save ec2 desc to %s"%fn_temp)
            with open(fn_temp, 'w') as fh:
                json.dump(operation_i, fh, default=json_serial, indent=4, sort_keys=True)


def get_cwDescAlarms(fn_dir, cloudwatch):
    get_cwCore(fn_dir, cloudwatch, 'describe_alarms', 'MetricAlarms')


def get_cwListMetrics(fn_dir, cloudwatch):
    get_cwCore(fn_dir, cloudwatch, 'list_metrics', 'Metrics')
    
    
def get_sns_listTopics(fn_dir, sns):
    # SNS paginator
    # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sns.html#SNS.Paginator.ListTopics
    PaginationConfig = {
        'MaxResults': 3000 # FIXME <<<<<<<<<<<<<<<<<<<<<<<<<<<
    }
    operation_name = 'list_topics'
    dict_key = 'Topics'
    paginator = sns.get_paginator(operation_name)
    for i, response in tqdm(enumerate(paginator.paginate(PaginationConfig=PaginationConfig)), desc="Downloading SNS topics list"):
        for j, operation_i in enumerate(response[dict_key]):
            # sys.stderr.write("%s/%s: %s\n"%(i, j, json.dumps(operation_i)))
            #return # FIXME
        
            tid = operation_i['TopicArn'].split(':')[-1]
            fn_temp = os.path.join(fn_dir, tid+'.json')
            with open(fn_temp, 'w') as fh:
                json.dump(operation_i, fh, default=json_serial, indent=4, sort_keys=True)
    
    
import datetime as dt
def get_cwGetMetricData(fn_dir1, cloudwatch, ec2):
    """
    cloudwatch - boto3 client for cloudwatch service
    ec2 - boto3 resource for ec2 service
    """
    
    # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudwatch.html#CloudWatch.Paginator.GetMetricData
    stat_list=[
        'Maximum',
        'Average',
        'Minimum',
        'SampleCount'
    ]
    metric_list = [
        'CPUUtilization'
    ]
    # utility variable so that all recorded data at this point get marked with the same timestamp
    dt_now_d = dt.datetime.now()
    dt_now_s = str(dt_now_d)
    seconds_in_one_day = 60*60*24 # 86400  # used for granularity (daily)
    seconds_in_one_hour = 60*60
    # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudwatch.html#CloudWatch.Paginator.DescribeAlarms
    PaginationConfig = {
        'MaxResults': 3000 # FIXME <<<<<<<<<<<<<<<<<<<<<<<<<<<
    }

    # if permissions were ok, continue
    # check ec2 permissions
    # iterate over all regions

    # ec2.instances is of class ec2.instancesCollectionManager
    # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/collections.html
    for instance_obj in tqdm(ec2.instances.all(), desc="Downloading AWS Cloudwatch metrics"):
        # collect daily data as well as hourly data
        for n_days, period in [
            (90, seconds_in_one_day),
            
            # hourly data
            # max n points is 1440, so cannot do 90 days with hourly in 1 shot
            # FIXME comment out since not used yet
            # (30, seconds_in_one_hour)
            ]:
        
            for stat_i in stat_list:
                for metric_i in metric_list:
                    
                    func_kwargs = dict(
                        MetricDataQueries = [
                            dict(
                                Id='m1', # ATM only 1 element in this list, so it doesn't matter
                                MetricStat=dict(
                                    Metric=dict(
                                        Namespace='AWS/EC2',
                                        MetricName=metric_i,
                                        Dimensions=[
                                            {
                                                'Name': 'InstanceId',
                                                'Value': instance_obj.id
                                            }
                                        ]
                                    ),
                                    Period=period,
                                    Stat=stat_i,
                                    Unit='Percent'
                                )
                            )
                        ],
                        StartTime=dt_now_d - dt.timedelta(days=n_days),
                        EndTime=dt_now_d
                    )
                    
                    # a description in simple english of the function arguments passed
                    desc_p1 = "%i seconds"%period
                    if period==seconds_in_one_day : desc_p1 = 'daily'
                    if period==seconds_in_one_hour: desc_p1 = 'hourly'
                    args_desc = '%s - %s - %i days - %s'%(metric_i, desc_p1, n_days, stat_i)

                    # path in which to save file
                    fn_dir2 = os.path.join(fn_dir1, args_desc)
                    os.makedirs(fn_dir2, exist_ok=True)
                    open(os.path.join(fn_dir2, '.gitkeep'), 'w').write('')
                    
                    # cannot use get_cwCore since the usage is different than describe-alarms and list-metrics                
                    operation_name='get_metric_data'
                    dict_key = 'MetricDataResults'
                    paginator = cloudwatch.get_paginator(operation_name)
                    for i, response in enumerate(paginator.paginate(PaginationConfig=PaginationConfig, **func_kwargs)):
                        for j, operation_i in enumerate(response[dict_key]):
                            fn_temp = os.path.join(fn_dir2, instance_obj.id+'.json')
                            with open(fn_temp, 'w') as fh:
                                json.dump(operation_i, fh, default=json_serial, indent=4, sort_keys=True)
              
              
"""
Just moving this out of ec2op for later integration
Check https://gitlab.com/autofitcloud/git-remote-aws/issues/4

def cw_getMetricStatistics(instance_id, cloudwatch):

    # utility variable so that all recorded data at this point get marked with the same timestamp
    dt_now_d = dt.datetime.now()
    dt_now_s = str(dt_now_d)
    seconds_in_one_day = 60*60*24 # 86400  # used for granularity (daily)
    seconds_in_one_hour = 60*60

    # if permissions were ok, continue
    # check ec2 permissions
    # iterate over all regions

    # collect daily data as well as hourly data
    for n_days, period in [
        (90, seconds_in_one_day),
        (30, seconds_in_one_hour) # max n points is 1440, so cannot do 90 days with hourly in 1 shot
        ]:
        func_args_1 = dict(
            Namespace='AWS/EC2',
            Dimensions=[
                {
                    'Name': 'InstanceId',
                    'Value': instance_id
                }
            ],
            MetricName='CPUUtilization',
            StartTime=dt_now_d - dt.timedelta(days=n_days),
            EndTime=dt_now_d,
            Period=period,
            Statistics=[
                'Maximum',
                'Average',
                'Minimum',
                'SampleCount'
            ],
            Unit='Percent'
        )
        desc_p1 = "(every %s seconds)"%period
        if period==seconds_in_one_day : desc_p1 = 'daily'
        if period==seconds_in_one_hour: desc_p1 = 'hourly'
        args_desc = 'CPU %s for %i days'%(desc_p1, n_days) # a description in simple english of the function arguments passed
        #print("processing", args_desc)

        response_1 = cloudwatch.get_metric_statistics(**func_args_1)

        response_2 = {
            'InstanceId': instance_id,
            'ArgsDesc': args_desc,
            'FetchTimestamp': dt_now_s,
            'CloudwatchMetricArgs': func_args_1,
            'CloudwatchMetricStatistics': response_1
        }

        yield response_2, args_desc
"""
                
                
EC2INSTANCESINFO = 'http://www.ec2instances.info/instances.json'
def get_awsCat(fn, ec2catalog=None):
    ec2catalog = EC2INSTANCESINFO
    logger.info("getting aws catalog from %s"%ec2catalog)

    # non-cached
    # https://3.python-requests.org/
    # from requests import HTTPSession
    # http = HTTPSession()
    # r = http.request('get', ec2catalog)

    # cached https://cachecontrol.readthedocs.io/en/latest/
    sess = requests.session()
    cached_sess = CacheControl(sess)
    r = cached_sess.request('get', ec2catalog)

    df_json = json.dumps(r.json(), indent=4, sort_keys=True)

    # prep save
    #logger.debug("mkdir %s"%fn['awsCat'])
    os.makedirs(fn['awsCat'], exist_ok=True)

    # save raw
    fn_temp = os.path.join(fn['awsCat'], 't0_raw.json')
    with open(fn_temp, 'w') as fh:
        fh.write(df_json)
        

    # Convert json to pandas dataframe
    # Test live with
    # >>> import pandas as pd, json
    # >>> df = pd.DataFrame(json.load(open("core/www.ec2instances.info/t0_raw.json")))
    # >>> df.head()
    #
    # df_pd = pd.read_json(r.json())
    df_pd = pd.DataFrame(r.json()) # [:2]

    # Gather prices from different regions into a list and calculate the average
    # remember that default_priceRegion goes to the pandas dataframe index, and the output is a dict, so all keys after that go to the python dict
    df_pd['price_list'] = df_pd['pricing'].apply(lambda x: np.array([float(x[y]['linux']['ondemand']) for y in x.keys() if 'linux' in x[y]]))
    df_pd['price_avg'] = df_pd['price_list'].apply(lambda x: None if len(x)==0 else np.average(x))
    
    # check that the prices in all the regions do not differ by more than 20%
    # In fact, there are plenty of price differences > 20%
    # eg 'c5d.xlarge' min/max are 0.192/0.252
    # So won't check this anymore
    # df_pd['price_min'] = df_pd['price_list'].apply(lambda x: None if len(x)==0 else x.min())
    # df_pd['price_max'] = df_pd['price_list'].apply(lambda x: None if len(x)==0 else x.max())
    # df_pd['price_diff'] = (df_pd['price_max'] - df_pd['price_min'])/df_pd['price_min']*100
    #if (df_pd['price_diff'] > 20).any():
    #    logger.warning("Found instance types with price difference within regions > 20%%: %s. Aborting"%(', '.join(df_pd.instance_type[df_pd['price_diff'] > 20].values)))

    # logger.debug("df_pd.head")
    # logger.debug(df_pd[['price_min', 'price_avg', 'price_max']].head())
    
    # replace the "pricing" field from a nested dict with the average
    # (from colab/20190624-crow/t1-raw aws)
    # This is for back-compatibility since my code expects pricing to be a single float
    df_pd['pricing'] = df_pd['price_avg']

    # rename columns as needed
    df_pd = df_pd.rename(columns={
        'instance_type': 'API Name',
        'vCPU': 'vCPUs',
        'memory': 'Memory',
        'storage': 'Instance Storage',
        'network_performance': 'Network Performance',
        'pricing': 'Linux On Demand cost',
    })

    # proceed
    df_pd = df_pd[['API Name', 'vCPUs', 'Memory', 'Instance Storage', 'Network Performance', 'Linux On Demand cost']]

    df_pd['family_l1'] = df_pd['API Name'].str.split('.').apply(lambda x: x[0][0])
    df_pd['family_l2'] = df_pd['API Name'].str.split('.').apply(lambda x: x[0])
    #df_pd['vCPUs'] = df_pd['vCPUs'].str.split(' ').apply(lambda x: x[0]).astype(int)
    #df_pd['Memory'] = df_pd['Memory'].str.split(' ').apply(lambda x: x[0]).astype(float)
    #df_pd['Linux On Demand cost'] = ( df_pd['Linux On Demand cost']
    #                                   .str.replace('unavailable', '')
    #                                   .str.split(' ')
    #                                   .apply(lambda x: x[0])
    #                                   .str.replace('$','')
    #                                   .map(lambda x: None if x=='' else float(x))
    #                             )

    df_pd = df_pd.sort_values(['family_l1', 'family_l2', 'vCPUs', 'Memory'])
    # df_pd.set_index(['family_l1', 'family_l2', 'API Name'], inplace=True)
    #df_pd = df_pd.sort_index()

    # df_json = df_pd.reset_index().to_json(orient='split')
    df_json = df_pd.to_json(orient='split')

    # pretty-print
    df_json = json.dumps(json.loads(df_json), indent=4, sort_keys=True)

    # save
    fn_temp = os.path.join(fn['awsCat'], 't1_processed.json')
    with open(fn_temp, 'w') as fh:
        fh.write(df_json)

    # append smaller types - family_l1
    df_l1 = df_pd.copy()
    df_l1 = df_l1.sort_values(['family_l1', 'vCPUs', 'Memory', 'family_l2']) # notice the sorting fields
    df_l1['type_smaller'] = df_l1.groupby('family_l1')['API Name'].shift(+1)
    df_l1 = df_l1.merge(
        df_l1[['API Name', 'vCPUs', 'Memory', 'Linux On Demand cost']].rename(columns={'API Name': 'type_smaller'}),
        on='type_smaller',
        how='outer',
        suffixes=['', '_smaller']
    )
    df_l1 = df_l1.sort_values(['family_l1', 'family_l2', 'vCPUs', 'Memory'])
    df_l1.set_index(['family_l1','family_l2','API Name'], inplace=True)

    df_l1['ratio_cpu'] = df_l1.vCPUs / df_l1.vCPUs_smaller
    df_l1['ratio_mem'] = df_l1.Memory / df_l1.Memory_smaller
    # df_l1['ratio_min'] = df_l1[['ratio_cpu', 'ratio_mem']].min(axis=1)
    df_l1['ratio_max'] = df_l1[['ratio_cpu', 'ratio_mem']].max(axis=1) # this is the bottleneck
    dfl1_json = df_l1.reset_index().to_json(orient='split')
    dfl1_json = json.dumps(json.loads(dfl1_json), indent=4, sort_keys=True) # pretty-print

    # save
    fn_temp = os.path.join(fn['awsCat'], 't3a_smaller_familyL1.json')
    with open(fn_temp, 'w') as fh:
        fh.write(dfl1_json)

    # append smaller types - family_l2
    df_l2 = df_pd.reset_index().copy()
    df_l2 = df_l2.sort_values(['family_l1', 'family_l2', 'vCPUs', 'Memory']) # notice the sorting fields
    df_l2['type_smaller'] = df_l2.groupby('family_l2')['API Name'].shift(+1) # notice the groupby field
    df_l2 = df_l2.merge(
        df_l2[['API Name', 'vCPUs', 'Memory', 'Linux On Demand cost']].rename(columns={'API Name': 'type_smaller'}),
        on='type_smaller',
        how='outer',
        suffixes=['', '_smaller']
    )
    df_l2 = df_l2.sort_values(['family_l1', 'family_l2', 'vCPUs', 'Memory'])
    df_l2.set_index(['family_l1','family_l2','API Name'], inplace=True)

    df_l2['ratio_cpu'] = df_l2.vCPUs / df_l2.vCPUs_smaller
    df_l2['ratio_mem'] = df_l2.Memory / df_l2.Memory_smaller
    # df_l2['ratio_min'] = df_l2[['ratio_cpu', 'ratio_mem']].min(axis=1)
    df_l2['ratio_max'] = df_l2[['ratio_cpu', 'ratio_mem']].max(axis=1) # this is the bottleneck
    dfl2_json = df_l2.reset_index().to_json(orient='split')
    dfl2_json = json.dumps(json.loads(dfl2_json), indent=4, sort_keys=True) # pretty-print

    # save
    fn_temp = os.path.join(fn['awsCat'], 't3b_smaller_familyL2.json')
    with open(fn_temp, 'w') as fh:
        fh.write(dfl2_json)



